src/coreComponents/constitutive/fluid/CubicEOSPhaseModel.hpp

```diff
@@ -0,0 +1,458 @@
+    {
+      aMixtureCoefficient += ( composition[ic] * composition[jc] * ( 1.0 - binaryInteractionCoefficients ) * sqrt( aPureCoefficient[ic] * aPureCoefficient[jc] ) );
+    }
+    bMixtureCoefficient += composition[ic] * bPureCoefficient[ic];
+  }
+}
+
+template< typename EOS_TYPE >
+GEOS_HOST_DEVICE
+void
+CubicEOSPhaseModel< EOS_TYPE >::
+computeCompressibilityFactor( integer const numComps,
+                              arrayView1d< real64 const > const composition,
+                              real64 const & binaryInteractionCoefficients,
+                              arraySlice1d< real64 const > const aPureCoefficient,
+                              arraySlice1d< real64 const > const bPureCoefficient,
+                              real64 const & aMixtureCoefficient,
+                              real64 const & bMixtureCoefficient,
+                              real64 & compressibilityFactor )
+{
+  // a Z3 + b Z2 + cZ + d = 0
+  real64 const a = 1.0;
+  real64 const b = ( EOS_TYPE::delta1 + EOS_TYPE::delta2 - 1.0 ) * bMixtureCoefficient - 1.0;
+  real64 const c = aMixtureCoefficient + EOS_TYPE::delta1 * EOS_TYPE::delta2 * bMixtureCoefficient * bMixtureCoefficient
+                   - ( EOS_TYPE::delta1 + EOS_TYPE::delta2 ) * bMixtureCoefficient * ( bMixtureCoefficient + 1.0 );
+  real64 const d = -( aMixtureCoefficient * bMixtureCoefficient
+                      + EOS_TYPE::delta1 * EOS_TYPE::delta2 * bMixtureCoefficient * bMixtureCoefficient * ( bMixtureCoefficient + 1.0 ) );
+
+  real64 roots[3]{};
+  integer numRoots = 0;
+  solveCubicPolynomial( a, b, c, d, roots, numRoots );
+
+  if( numRoots == 1 )
+  {
+    compressibilityFactor = roots[0];
+  }
+  else
+  {
+    real64 zMin =  LvArray::NumericLimits< real64 >::max;
+    real64 zMax = -LvArray::NumericLimits< real64 >::max;
+
+    for( integer i = 0; i < 3; ++i )
+    {
+      // skip unphysical roots
+      if( roots[i] > bMixtureCoefficient )
+      {
+        // choose the root according to Gibbs' free energy minimization
+        if( zMin < roots[i] )
+        {
+          zMin = roots[i];
+        }
+        if( zMax > roots[i] )
+        {
+          zMax = roots[i];
+        }
+      }
+    }
+
+    stackArray1d< real64, maxNumComps > logFugacityCoefficientsMax( numComps );
+    stackArray1d< real64, maxNumComps > logFugacityCoefficientsMin( numComps );
+    computeLogFugacityCoefficients( numComps, composition, binaryInteractionCoefficients, zMin,
+                                    aPureCoefficient, bPureCoefficient, aMixtureCoefficient, bMixtureCoefficient,
+                                    logFugacityCoefficientsMin.toSlice() );
+    computeLogFugacityCoefficients( numComps, composition, binaryInteractionCoefficients, zMax,
+                                    aPureCoefficient, bPureCoefficient, aMixtureCoefficient, bMixtureCoefficient,
+                                    logFugacityCoefficientsMin.toSlice() );
```

Comment:
Second call should pass `logFugacityCoefficientsMax.toSlice()`

Reply:
Done